cloud-exec.md

```diff
@@ -0,0 +1,182 @@
+
+Multiple cloud instance optimization technicques might be used to optimize
+  the usage of resources or execution time:
+1. **Spot instances.**
+2. **Transparent spot instance.** Recover the execution if a spot instance
+   was terminated. DVC checkpoints and pipeline stages should be used for
+   preserving the state.
+3. **Volumes.** Volumes can be attached and reused in instances to minimize
+   data cache synchronization time.
+4. **Shared volumes.** Separate cloud services (such as Multi-Attach EBS or
+   EFS) might be needed for sharing data cache between multiple instances.
+
+## Monitoring 
+
+Additional features might be needed to improve user experienc:
+1. **State monitoring.** A user should see if an instance was already
+   allocated, what stage is running.
+2. **Metrics monitoring.**  Latest metrics monitoring (with some reasonable
+   delay - 10-30-60 seconds)
+
+
+## Resource orchestrators
+
+1. AWS
+2. Azure
+3. GCP (optional)
+4. Remote over SSH
+5. Kubernetes (K8S)
+6. HashiCorp Nomad (optional)
+7. Container services (ECS and Azure/GCP analogs)
+
+
+# Detailed design
+
+## Iterative Terraform provider
+
+The cloud management part seems like a heavy part of the project (from
+  the implementation point of view). However, the problem of executing and
+  managing cloud instances was solved in the CML project.
+
+CML separates out the user API part (commands) and the cloud instace
+  management code part.
+The cloud management is extracted to a separate **Iterative Terraform (TF)
+  provider**:
+  https://github.com/iterative/terraform-provider-iterative
+
+The Iterative provider is specific to data scientists, not DevOps (TF's
+  target audience). It has a bit simplified API and some ML specific
+  functionality (some of these are coming) like: recover spot-instance if
+  it was deleted or attach a reusable drive.
+
+The Iteratve TF provider can be used from DVC for cloud instace management.
+Some Python libraries might be required  for access to GoLang specific
+terraform-provider.
+
+Today the provider supports:
+- AWS, Azure, K8S orchestrators
+- Spot instances and Volumes optimizations.
+A few of the initial scenarios can be already implemented with these
+functionalities. More functionality is comming.
+
+## Instances configuration
+
+Instance configuration is a definition of an instance and its behavior:
+- orchestrator: AWS, GCP, Azure, remote, K8S, ...
+- instance type
+- volume
+- many other options
+
+Instance configuration is well developed in CML project and can be reused.
+
+## Executors and naming
+
+Ideally, a user should be able to define or overwrite executors in config
+(`dvc.yml` or `.dvc/config`). `dvc exp run` should execute,
+provision/terminate if needed.
+
+The executor definition (`p3.large` AWS instance in `us-west` with `abcdef`
+volume attached) should be decoupled from pipeline definition the same way
+as remotes are: stage `train` runs on `my-gpu-tesla`, not executor
+definition with `p3.large`.
```

Comment:
If we define in `dvc.yaml`, it makes it hard to run elsewhere (like locally), and it couples experiment history and infrastructure (similar to problems with changing remotes today). Maybe having a local config option or having an `exp run --executor` flag is sufficient flexibility here? What would we do differently with remotes if we were starting from scratch?